src/model/guild/automod.rs

```diff
@@ -0,0 +1,666 @@
+                        Ok(Action::Timeout(Duration::from_secs(timeout.duration)))
+                    },
+                    ActionType::Unknown(unknown) => Ok(Action::Unknown(unknown)),
+                }
+            }
+        }
+
+        deserializer.deserialize_any(Visitor)
+    }
+}
+
+impl Serialize for Action {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let has_metadata = matches!(self, Self::Alert(_) | Self::Timeout(_));
+
+        let len = 1 + usize::from(has_metadata);
+        let mut s = serializer.serialize_struct("Action", len)?;
+
+        s.serialize_field("type", &self.kind())?;
+        match *self {
+            Self::Alert(channel_id) => {
+                s.serialize_field("metadata", &Alert {
+                    channel_id,
+                })?;
+            },
+            Self::Timeout(duration) => {
+                s.serialize_field("metadata", &Timeout {
+                    duration: duration.as_secs(),
+                })?;
+            },
+            _ => {},
+        }
+
+        s.end()
+    }
+}
+
+impl Action {
+    #[must_use]
+    pub fn kind(&self) -> ActionType {
+        match self {
+            Self::BlockMessage => ActionType::BlockMessage,
+            Self::Alert(_) => ActionType::Alert,
+            Self::Timeout(_) => ActionType::Timeout,
+            Self::Unknown(unknown) => ActionType::Unknown(*unknown),
+        }
+    }
+}
+
+/// Type of [`Action`].
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Deserialize, Serialize)]
+#[serde(from = "u8", into = "u8")]
+#[non_exhaustive]
+pub enum ActionType {
+    /// Blocks the content of a message according to the rule.
+    BlockMessage,
+    /// Logs user content to a specified channel.
+    Alert,
+    /// Timeout user for a specified duration.
+    ///
+    /// A `Timeout` action can only be setup for [`Keyword`] rules.
+    /// [`Permissions::MODERATE_MEMBERS`] permission is required to use the `Timeout` action type.
+    ///
+    /// [`Keyword`]: TriggerType::Keyword
+    /// [`Permissions::MODERATE_MEMBERS`]: crate::model::Permissions::MODERATE_MEMBERS
+    Timeout,
+    Unknown(u8),
+}
+
+impl From<u8> for ActionType {
+    fn from(value: u8) -> Self {
+        match value {
+            1 => Self::BlockMessage,
+            2 => Self::Alert,
+            3 => Self::Timeout,
+            unknown => Self::Unknown(unknown),
+        }
+    }
+}
+
+impl From<ActionType> for u8 {
+    fn from(value: ActionType) -> Self {
+        match value {
+            ActionType::BlockMessage => 1,
+            ActionType::Alert => 2,
+            ActionType::Timeout => 3,
+            ActionType::Unknown(unknown) => unknown,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use serde_test::Token;
+
+    use super::*;
+
+    #[test]
+    fn rule_trigger_serde() {
+        #[derive(Debug, PartialEq, Deserialize, Serialize)]
+        struct Rule {
```

Comment:
Btw why redefine Rule for the test instead of using the existing Rule struct?

Reply:
This saves a bunch of tokens.